pandas/_libs/tslibs/strptime.pyx

```diff
@@ -185,9 +195,55 @@ def array_strptime(
         elif is_datetime64_object(val):
             iresult[i] = get_datetime64_nanos(val, NPY_FR_ns)
             continue
+        elif (
+                (is_integer_object(val) or is_float_object(val))
+                and (val != val or val == NPY_NAT)
+        ):
+            iresult[i] = NPY_NAT
+            continue
         else:
             val = str(val)
 
+        if (iso_format and not (fmt == "%Y%m%d" and len(val) != 8)):
+            # There is a fast-path for ISO8601-formatted strings.
+            # BUT for %Y%m%d, it only works if the string is 8-digits long.
+            string_to_dts_failed = string_to_dts(
+                val, &dts, &out_bestunit, &out_local,
+                &out_tzoffset, False, fmt, exact
+            )
+            if string_to_dts_failed:
+                # An error at this point is a _parsing_ error
+                # specifically _not_ OutOfBoundsDatetime
+                if is_coerce:
+                    iresult[i] = NPY_NAT
+                    continue
+                raise ValueError(
+                    f"time data \"{val}\" at position {i} doesn't "
+                    f"match format \"{fmt}\""
+                )
+            # No error reported by string_to_dts, pick back up
+            # where we left off
+            value = npy_datetimestruct_to_datetime(NPY_FR_ns, &dts)
+            if out_local == 1:
+                # Store the out_tzoffset in seconds
+                # since we store the total_seconds of
+                # dateutil.tz.tzoffset objects
+                # out_tzoffset_vals.add(out_tzoffset * 60.)
+                tz = timezone(timedelta(minutes=out_tzoffset))
+                result_timezone[i] = tz
+                # value = tz_localize_to_utc_single(value, tz)
+                out_local = 0
+                out_tzoffset = 0
+            iresult[i] = value
+            try:
+                check_dts_bounds(&dts)
+            except ValueError:
+                if is_coerce:
+                    iresult[i] = NPY_NAT
+                    continue
+                raise
+            continue
```

Comment:
this pretty-much matches 

https://github.com/pandas-dev/pandas/blob/65987973b4299004aa794f84ef187a1b6b58aec1/pandas/_libs/tslib.pyx#L598-L668

but it's simpler as we don't need to try `parse_datetime_string`. That's because if we got here, we know that we're expecting some specific ISO8601 format, so if `string_to_dts` can't parse it, then we need to coerce/raise/ignore, but there's no need to try other formats